plugins/catalog-backend-module-msgraph/src/microsoftGraph/read.ts

```diff
@@ -19,255 +19,138 @@ import {
     options.queryMode,
-  )) {
-    // Process all users in parallel, otherwise it can take quite some time
-    promises.push(
-      limiter(async () => {
-        let userPhoto;
-        try {
-          userPhoto = await client.getUserPhotoWithSizeLimit(
-            user.id!,
-            // We are limiting the photo size, as users with full resolution photos
-            // can make the Backstage API slow
-            120,
-          );
-        } catch (e) {
-          options.logger.warn(`Unable to load photo for ${user.id}, ${e}`);
-        }
-
-        const entity = await transformer(user, userPhoto);
-
-        if (!entity) {
-          return;
-        }
-
-        users.push(entity);
-      }),
-    );
-  }
-
-  // Wait for all users and photos to be downloaded
-  await Promise.all(promises);
+  );
 
-  return { users };
+  return {
+    users: await transformUsers(
+      client,
+      users,
+      options.logger,
+      options.transformer,
+    ),
+  };
 }
 
 export async function readMicrosoftGraphUsersInGroups(
   client: MicrosoftGraphClient,
   options: {
-    queryMode?: 'basic' | 'advanced';
+    queryMode?: QueryMode;
     userExpand?: string;
     userSelect?: string[];
     userGroupMemberSearch?: string;
     userGroupMemberFilter?: string;
     groupExpand?: string;
     transformer?: UserTransformer;
-    logger: Logger;
+    logger?: Logger;
   },
 ): Promise<{
   users: UserEntity[]; // With all relations empty
 }> {
-  const users: UserEntity[] = [];
-
   const limiter = limiterFactory(10);
 
-  const transformer = options.transformer ?? defaultUserTransformer;
   const userGroupMemberPromises: Promise<void>[] = [];
-  const userPromises: Promise<void>[] = [];
+  const userGroupMembers = new Map<string, User>();
 
-  const groupMemberUsers: Set<string> = new Set();
+  options.logger?.info('Loading users by group membership');
 
   for await (const group of client.getGroups(
     {
       expand: options.groupExpand,
       search: options.userGroupMemberSearch,
       filter: options.userGroupMemberFilter,
+      select: ['id', 'displayName'],
+      top: PAGE_SIZE,
     },
     options.queryMode,
   )) {
     // Process all groups in parallel, otherwise it can take quite some time
     userGroupMemberPromises.push(
       limiter(async () => {
-        for await (const member of client.getGroupMembers(group.id!)) {
-          if (!member.id) {
-            continue;
-          }
-
-          if (member['@odata.type'] === '#microsoft.graph.user') {
-            groupMemberUsers.add(member.id);
-          }
+        let groupMemberCount = 0;
+        for await (const user of client.getGroupUserMembers(
+          group.id!,
+          {
+            expand: options.userExpand,
+            top: PAGE_SIZE,
+          },
+          options.queryMode,
+        )) {
+          userGroupMembers.set(user.id!, user);
+          groupMemberCount++;
         }
+        options.logger?.info('Read users from group', {
```

Comment:
This is going to spew a VERY large amount of logs on large orgs won't it? Feels like this should be at debug or trace level or skipped

Reply:
Yes, but it also gives some indication of progress.  For a medium to large sized org, the ms graph ingestion process can take many many hours, without logging any signs of progress.  That leads to us getting a lot of support requests on github & discord about "WHy is it not working", the answer to which almost always turns out to be "Give it more time" or "Narrow your query to be more specific and load less data".

Reply:
Thinking about this more, think I'm happy to demote this to debug level.  It may slightly harm the onboarding experience by not so clearly showing hte progress being made, but that shouldn't be a problem once you're set up.  Asking on boarders to enable debug logging for more details doesn't seem unreasonable.

Reply:
Fixed in #15891